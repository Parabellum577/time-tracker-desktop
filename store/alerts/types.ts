import { ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IError } from '@services/types'

export type AlertsAction = ActionType<typeof actions>

export interface IAlert {
  alertType: 'success' | 'error' | 'warning' | 'coldstart'
  alertMessage: string | IError
}

export interface IAlertsState {
  alertsArray: IAlert[]
}
