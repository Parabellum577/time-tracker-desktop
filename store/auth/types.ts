import { ActionType } from 'typesafe-actions'
import { DeepReadonly } from 'utility-types'

import * as actions from './actions'
import { IError } from '@types'

export type AuthAction = ActionType<typeof actions>

export type IAuthState = DeepReadonly<{
  readonly authorized: boolean
  readonly error?: IError | null
  readonly loading: boolean
  readonly isLoadingData: boolean
}>
