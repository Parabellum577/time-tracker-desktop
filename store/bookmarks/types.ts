import { IFullIssue } from '@types'

export interface IBookmarks {
  extendedBookmarks: IFullIssue[]
}
