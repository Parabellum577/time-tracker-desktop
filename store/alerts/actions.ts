import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
