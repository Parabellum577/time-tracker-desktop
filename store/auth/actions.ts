import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
