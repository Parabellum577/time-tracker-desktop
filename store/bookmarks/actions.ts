import { createAsyncAction, createAction } from 'typesafe-actions'
import { IError, IFullIssue } from '@types'

export const getBookmarks = createAsyncAction(
  'bookmarks/GET_BOOKMARKS_REQUEST',
  'bookmarks/GET_BOOKMARKS_SUCCESS',
  'bookmarks/GET_BOOKMARKS_FAILURE',
)<void, IFullIssue[], IError>()

export const handleBookmark = createAsyncAction(
  'bookmarks/HANDLE_BOOKMARK_REQUEST',
  'bookmarks/HANDLE_BOOKMARK_SUCCESS',
  'bookmarks/HANDLE_BOOKMARK_FAILURE',
)<IFullIssue, IFullIssue, IError>()

export const clearBookmarks = createAction('plannings/CLEAR_BOOKMARKS', action => () => action())
