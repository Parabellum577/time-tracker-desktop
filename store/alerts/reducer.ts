import { getType } from 'typesafe-actions'

import * as actions from './actions'
import { IAlertsState, AlertsAction } from './types'

const initialState: IAlertsState = {
  alertsArray: [],
}

export default (state: IAlertsState = initialState, action: AlertsAction) => {
  let messageToShow

  switch (action.type) {
    case getType(actions.showAlertMessage.success):
      const newAlert = action.payload

      switch (newAlert.alertMessage) {
        case 'NETWORK_ERROR':
          messageToShow = 'You are Offline. Please check your internet connection'
          break
        case 'MAIL_CONFIRMATION_EXPIRED':
          messageToShow = 'Your account has been expired. Please confirm your login'
          break
        case 'INVALID_LOGIN':
          messageToShow = 'Email is incorrect'
          break
        case 'INVALID_PASSWORD':
          messageToShow = 'Password is incorrect'
          break
        case 'USER_BANNED':
          messageToShow = 'Your account is banned'
          break
        case 'LOGIN_EXISTS':
          messageToShow = 'Account with this email is already exist'
          break
        case 'INVALID_TRACKER_URL':
          messageToShow = 'Integration with this URL is not exist'
          break
        case 'INVALID_CREDENTIALS':
          messageToShow = 'Integration credentials are not valid'
          break
        case 'CREDENTIALS_EXIST':
          messageToShow = 'Integration with these credentials is already connected'
          break
        case 'TRACKER_NOT_FOUND':
          messageToShow = 'Integration with this URL is not exist'
          break
        default:
          messageToShow = newAlert.alertMessage
          break
      }
      return {
        alertsArray: [...state.alertsArray, { ...newAlert, alertMessage: messageToShow }],
      }

    case getType(actions.clearAlertMessage.success):
      return {
        ...state,
        alertsArray: [...state.alertsArray.slice(1)],
      }

    default:
      return state
  }
}
