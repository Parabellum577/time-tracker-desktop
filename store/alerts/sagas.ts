import * as _ from 'lodash'
import { SagaIterator } from 'redux-saga'
import { put, take, select, spawn, delay } from 'redux-saga/effects'
import { getType } from 'typesafe-actions'

import { IRootState } from '@store/rootReducer'
import * as actions from './actions'
import { IAlert } from './types'

export const alertsArrayFromStore = (state: IRootState) => state.alerts.alertsArray

export function* showAlertsSaga(): SagaIterator {
  while (true) {
    const { payload } = yield take(getType(actions.showAlertMessage.request))
    let alertsArray: IAlert[] = yield select(alertsArrayFromStore)
    if (_.isEmpty(alertsArray)) {
      yield spawn(function*() {
        while (true) {
          yield delay(3000)
          alertsArray = yield select(alertsArrayFromStore)
          if (_.isEmpty(alertsArray)) {
            return
          }
          yield put(actions.clearAlertMessage.success())
        }
      })
    }
    yield put(actions.showAlertMessage.success(payload))
  }
}

export function* handleShowAlerts(alert: IAlert) {
  yield put(actions.showAlertMessage.success(alert))
}

export function* clearAlertsSaga(): SagaIterator {
  while (true) {
    yield take(getType(actions.clearAlertMessage.request))
    yield put(actions.clearAlertMessage.success())
  }
}
