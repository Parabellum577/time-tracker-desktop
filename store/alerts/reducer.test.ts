import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import { IAlertsState } from './types'
import alerts from './reducer'

import { SIMPLE_ALERT } from '@test-helpers/sample'

describe('[Store] Alerts Reducer', () => {
  it('[Success] [Snapshot] Should handle showAlertMessage', () => {
    const showAlertAction: ActionType<typeof actions> = {
      type: getType(actions.showAlertMessage.success),
      payload: { ...SIMPLE_ALERT },
    }

    const currentState: IAlertsState = {
      alertsArray: [{ ...SIMPLE_ALERT }],
    }

    expect(alerts(currentState, showAlertAction)).toMatchSnapshot()
  })

  it('[Failure] Should handle showAlertMessage', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.showAlertMessage.failure),
      payload: {
        code: 666,
        message: 'SOME_ERROR',
      },
    }

    const currentState: IAlertsState = {
      alertsArray: [],
    }

    expect(alerts(currentState, action)).toMatchSnapshot()
  })

  it('[Success] [Snapshot] Should handle clearAlertMessage', () => {
    const clearAlertAction: ActionType<typeof actions> = {
      type: getType(actions.clearAlertMessage.success),
      payload: null,
    }

    const currentState: IAlertsState = {
      alertsArray: [],
    }

    expect(alerts(currentState, clearAlertAction)).toMatchSnapshot()
  })

  it('[Failure] Should handle clearAlertMessage', () => {
    const clearAlertAction: ActionType<typeof actions> = {
      type: getType(actions.clearAlertMessage.failure),
      payload: {
        code: 666,
        message: 'SOME_ERROR',
      },
    }

    const currentState: IAlertsState = {
      alertsArray: [{ ...SIMPLE_ALERT }],
    }

    expect(alerts(currentState, clearAlertAction)).toMatchSnapshot()
  })
})
