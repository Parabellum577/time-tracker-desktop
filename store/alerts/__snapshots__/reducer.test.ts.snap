// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Alerts Reducer [Failure] Should handle clearAlertMessage 1`] = `
Object {
  "alertsArray": Array [
    Object {
      "alertMessage": "INVALID_LOGIN",
      "alertType": "error",
    },
  ],
}
`;

exports[`[Store] Alerts Reducer [Failure] Should handle showAlertMessage 1`] = `
Object {
  "alertsArray": Array [],
}
`;

exports[`[Store] Alerts Reducer [Success] [Snapshot] Should handle clearAlertMessage 1`] = `
Object {
  "alertsArray": Array [],
}
`;

exports[`[Store] Alerts Reducer [Success] [Snapshot] Should handle showAlertMessage 1`] = `
Object {
  "alertsArray": Array [
    Object {
      "alertMessage": "INVALID_LOGIN",
      "alertType": "error",
    },
    Object {
      "alertMessage": "Email is incorrect",
      "alertType": "error",
    },
  ],
}
`;
